pkg/envd/envdserver.go

```diff
@@ -75,20 +78,72 @@ func (e *envdServerEngine) CleanEnvdIfExists(ctx context.Context, name string, f
 }
 
 // StartEnvd creates the container for the given tag and container name.
-func (e *envdServerEngine) StartEnvd(ctx context.Context, tag, name, buildContext string,
-	gpuEnabled bool, numGPUs int, sshPort int, g ir.Graph, timeout time.Duration,
-	mountOptionsStr []string) (string, string, error) {
-	return "", "", errors.New("not implemented")
+func (e *envdServerEngine) StartEnvd(ctx context.Context, so StartOptions) (*StartResult, error) {
+	if so.EnvdServerSource == nil {
+		return nil, errors.New("failed to get the envd server specific options")
+	}
+
+	req := servertypes.EnvironmentCreateRequest{
+		IdentityToken: e.IdentityToken,
+		Image:         so.Image,
+	}
+
+	resp, err := e.EnvironmentCreate(ctx, req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create the environment")
+	}
+
+	if err := e.WaitUntilRunning(
+		ctx, resp.ID, so.Timeout); err != nil {
+		return nil, errors.Wrap(err, "failed to wait until the container is running")
+	}
+
+	result := &StartResult{
+		SSHPort: 2222,
+		Address: "",
+		Name:    resp.ID,
+	}
+	return result, nil
 }
 
 func (e *envdServerEngine) IsRunning(ctx context.Context, name string) (bool, error) {
-	return false, errors.New("not implemented")
+	req := servertypes.EnvironmentListRequest{
+		IdentityToken: e.IdentityToken,
+	}
+
+	resp, err := e.EnvironmentList(ctx, req)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to list the environment")
+	}
+	return resp.Pod.Status.Phase == v1.PodRunning, nil
 }
 
 func (e *envdServerEngine) Exists(ctx context.Context, name string) (bool, error) {
 	return false, errors.New("not implemented")
 }
 
 func (e *envdServerEngine) WaitUntilRunning(ctx context.Context, name string, timeout time.Duration) error {
-	return errors.New("not implemented")
+	logger := logrus.WithField("container", name)
+	logger.Debug("waiting to start")
+
+	// First, wait for the container to be marked as started.
+	ctxTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	for {
+		select {
+		case <-time.After(waitingInternal):
```

Comment:
Typo: `waitingInterval`.

Need to update `docker.go` file. I think it should be defined in `type.go` or `engine.go`.

Reply:
SGTM.